### PR TITLE
Scheduler should not stop one-shot replications

### PR DESCRIPTION
**Pull request by nickva**
Continuously starting and stopping one-shot (i.e. non-continuous) replications
alters their semantics -- replicating a particular snapshot of the source.
Changes which were made after replication has started, typically would not
be seen on the target. If replication crashes, or is otherwise restarted, that
constraint will be broken, but that those are exceptional cases.

Fix scheduler to treat one-shot replication differently, specifically:
- During job rotation, one-shot replication will stay running. And will not be
  stopped.
- However, pending one-shot replications don't get special treatment, they
  follow the basic algorithm of "job started last gets to run first".
- One example when one-shot replications will stop, is when user adjusts
  `max_jobs` parameter to a lower value. However, even in that case, first an
  attempt is made to evict all continous replications and only if that didn't
  satisfy the maximum constraint, some one-shot replication will be stopped.

Because this new behavior complicates scheduler behavior even more, add unit
tests to check various scheduler properties and edge cases.

### CODE REVIEW

File: src/couch_replicator_scheduler.erl [hunk 1]
```diff
@@ -50,6 +50,7 @@
 -record(job, {
           id :: job_id(),
           rep :: #rep{},
+          continuous :: boolean(),
```
Comment (rnewson) [inline src/couch_replicator_scheduler.erl]: this information is inside the `rep` field, I don't think we should make a copy.

File: src/couch_replicator_scheduler.erl [hunk 2]
```diff
@@ -369,9 +370,10 @@ running_job_count() ->
     ets:info(?MODULE, size) - pending_job_count().
 
 
--spec running_jobs() -> [#job{}].
-running_jobs() ->
-    ets:select(?MODULE, [{#job{pid = '$1', _='_'}, [{is_pid, '$1'}], ['$_']}]).
+-spec running_jobs(boolean()) -> [#job{}].
+running_jobs(Continuous) when Continuous =:= true ; Continuous =:= false->
```
Comment (rnewson) [inline src/couch_replicator_scheduler.erl]: this mingles together two concerns, needlessly complicated.

File: src/couch_replicator_scheduler.erl [hunk 3]
```diff
@@ -279,13 +281,12 @@ start_jobs(Count, State) ->
     [start_job_int(Job, State) || Job <- Runnable3],
     ok.
 
-
-stop_jobs(Count, State) ->
-    Running0 = running_jobs(),
+-spec stop_jobs(non_neg_integer(), boolean(), #state{}) -> non_neg_integer().
+stop_jobs(Count, Continuous, State) ->
+    Running0 = running_jobs(Continuous),
     Running1 = lists:sort(fun oldest_job_first/2, Running0),
```
Comment (rnewson) [inline src/couch_replicator_scheduler.erl]: `Running1 = lists:filter(fun not_continuous/1, Running0),` and adjust the variable names further down.

Comment (rnewson) [inline src/couch_replicator_scheduler.erl]: that is, let `running_jobs/0` return all running jobs, then filter out the one-shot replications, then sort, then truncate. I get that it seems inefficient, but it makes for simple and focused functions.

File: src/couch_replicator_scheduler.erl [hunk 4]
```diff
@@ -451,10 +453,19 @@ reschedule(State) ->
...
 stop_excess_jobs(State, Running) ->
     #state{max_jobs=MaxJobs} = State,
-    if Running > MaxJobs ->
-        stop_jobs(Running - MaxJobs, State);
+    StopCount = Running - MaxJobs,
+    if StopCount > 0 ->
+        Stopped = stop_jobs(StopCount, true, State),
+        OneshotLeft = StopCount - Stopped,
+        if OneshotLeft > 0 ->
+            stop_jobs(OneshotLeft, false, State),
```
Comment (rnewson) [inline src/couch_replicator_scheduler.erl:464]: doesn't this violate the point of the patch? (that we won't stop one-shot replications).

File: src/couch_replicator_scheduler.erl (hunk 3 above)
Comment (nickva) [inline src/couch_replicator_scheduler.erl]: Agree. Good point, will fix.

File: src/couch_replicator_scheduler.erl (hunk 1 above)
Comment (nickva) [inline src/couch_replicator_scheduler.erl]: This was mainly to make using a select pattern easier further below. 

`ets:select(?MODULE, [{#job{pid = '$1', continuous = Continuous, _='_'},` 

However, I like the idea of pulling out filtering of continuous to a separate function (as indicated below, so then this doesn't apply anymore). 

Will fix it.

File: src/couch_replicator_scheduler.erl (hunk 4 above)
Comment (nickva) [inline src/couch_replicator_scheduler.erl:464]: I think it is more clear how scheduler works if `max_jobs` really means `max_jobs`. This stopping code will only run after user (an operator) explicitly adjusts max_jobs to a lower value. Presumably they have some good reasons to do so. If this doesn't stop the jobs, there might be a need for a `really_max_jobs` setting in the future.

Also if we break the "Running <= MaxJobs"  constraint would mean changing the algorithm a bit,  as `rotate` function relies on it `if Running == MaxJobs, Pending > 0 -> .. start/stop ...`

File: src/couch_replicator_scheduler.erl (hunk 2 above)
Comment (nickva) [inline src/couch_replicator_scheduler.erl]: Yap, makes sense. This should be a separate filtering function.

File: src/couch_replicator_scheduler.erl [hunk 5]
```diff
@@ -280,12 +280,14 @@ start_jobs(Count, State) ->
...
 
-stop_jobs(Count, State) ->
+-spec stop_jobs(non_neg_integer(), boolean(), #state{}) -> non_neg_integer().
+stop_jobs(Count, Continuous, State) ->
     Running0 = running_jobs(),
-    Running1 = lists:sort(fun oldest_job_first/2, Running0),
-    Running2 = lists:sublist(Running1, Count),
-    [stop_job_int(Job, State) || Job <- Running2],
-    ok.
+    ContinuousPred = fun(Job) -> is_continuous(Job) =:= Continuous end,
```
Comment (rnewson) [inline src/couch_replicator_scheduler.erl]: make this a function, like the others.

File: src/couch_replicator_scheduler.erl (hunk 4 above)
Comment (rnewson) [inline src/couch_replicator_scheduler.erl:464]: agreed.

File: src/couch_replicator_scheduler.erl [hunk 6]
```diff
@@ -280,12 +280,14 @@ start_jobs(Count, State) ->
     ok.
 
 
-stop_jobs(Count, State) ->
+-spec stop_jobs(non_neg_integer(), boolean(), #state{}) -> non_neg_integer().
+stop_jobs(Count, Continuous, State) ->
```
Comment (rnewson) [inline src/couch_replicator_scheduler.erl]: I suggest `IfContinuous` for clarity.

Comment (rnewson) [inline src/couch_replicator_scheduler.erl]: or `IsContinuous` to match fun

**Comment by rnewson** [comment]
+1
